### PR TITLE
Add ADS support for ECDS

### PR DESCRIPTION
**Pull request by delong-coder**
This adds support for setting ECDS resources in the ADS.

Note that ECDS is not supported in V2, only changed V3 server code.

Related issue: https://github.com/envoyproxy/go-control-plane/issues/398

To add ECDS, I have to change the pkg/server, I'm not sure why it can't be modified.

Please help review. Thanks!

### CODE REVIEW

**Comment by fxposter** [comment]
1. v3 is generated from v2, so if you change v3 directly - it will be overwritten by running ./create_version.sh
2. given the fact that v2 and v3 have a generic handling of all resources and special-casing only a handful of them, do you really need to special-case this one? ie: if you just remove your code from server.go it should work out of the box.

**Comment by delong-coder** [comment]
> 1. v3 is generated from v2, so if you change v3 directly - it will be overwritten by running ./create_version.sh
> 2. given the fact that v2 and v3 have a generic handling of all resources and special-casing only a handful of them, do you really need to special-case this one? ie: if you just remove your code from server.go it should work out of the box.

Sure, I got your point, I will change the v2 code, then let ./create_version.sh generate the v3 code. Thanks.

**Comment by delong-coder** [comment]
Sorry, @fxposter , I tried adding my code into the v2 server, but it didn't work, because the ECDS is only enabled in V3, it needs the v3 discovery server.
I can't compile the ECDS with server v2 code unless I add v3 rest and sotw server to our /pkg/server/v2/server.go, that would be super ugly.
So maybe it's time to separate those 2 versions, stop creating v3 from v2 code, otherwise, we may never support the ECDS. 
@snowp  @kyessenov @jyotimahapatra

**Comment by delong-coder** [comment]
Thanks @peergynt ! 
So I think it's a common pain for lacking supporting ECDS.
Shall we go forward?

File: pkg/server/sotw/v3/server.go
Comment (fxposter) [inline pkg/server/sotw/v3/server.go:98]: If you remove changes in this file everything will still work out of the box, isn't it?

Comment (delong-coder) [reply pkg/server/sotw/v3/server.go:98]: Yes, the ECDS will still work, but we can't use it with the ADS server.

Comment (delong-coder) [reply pkg/server/sotw/v3/server.go:98]: I mean pkg/server/v3/server.go, not quite sure about sotw server.

Comment (fxposter) [reply pkg/server/sotw/v3/server.go:98]: yes, I am talking about this specific file :) maybe these can go via generic `responses`: https://github.com/envoyproxy/go-control-plane/pull/400/files#diff-0173d7909b484a388922cc5438a5b94b78b72fe780a531a0ed246bfa940a69f8R101

All other files are actually needed and the problem is - with current build setup it cannot be merged

Comment (delong-coder) [reply pkg/server/sotw/v3/server.go:98]: Not sure lacking extension config watcher has other potential effects or not.

Yes, other files are needed, can I change the current build setup and stop copy those files from v2 code?
I think that should be approved by someone.

Comment (fxposter) [reply pkg/server/sotw/v3/server.go:98]: I think nowadays it's up to @kyessenov maybe?

**Comment by snowp** [comment]
FYI this should be much easier to land once we land #412

**Comment by delong-coder** [comment]
> FYI this should be much easier to land once we land #412

Sounds good!

**Comment by alecholmez** [comment]
@delong-coder Has this been fixed for the v2 removal?

**Comment by alecholmez** [comment]
@snowp I think this is good? Just another type for SOTW

File: pkg/test/resource/v3/resource.go
Comment (alecholmez) [inline pkg/test/resource/v3/resource.go:309]: Can we move away from ptypes? It's been deprecated. I believe the replace here is:

```
anypb.New()
```

File: pkg/server/v3/server_test.go
Comment (alecholmez) [inline pkg/server/v3/server_test.go:381]: Thanks for the catch :)

**Review by alecholmez — CHANGES_REQUESTED** [review]
@snowp This addressed ADS support for ECDS, but I believe if we move to my refactor of SOTW to use the generic responses paradigm then we don't have to keep adding new hardcoded types for ADS support. Though I'm not entirely sure if the extra go routine overhead is worth it. Would it be worth comparing that refactor vs what we have now with a benchmark?

**Comment by snowp** [comment]
>Would it be worth comparing that refactor vs what we have now with a benchmark?

Yeah, I think that would be helpful

File: pkg/server/v3/server_test.go
Comment (delong-coder) [reply pkg/server/v3/server_test.go:379]: No problem.

File: pkg/test/resource/v3/resource.go
Comment (delong-coder) [reply pkg/test/resource/v3/resource.go:332]: I see `ptypes` is used everywhere in this file. Shall we change all the `ptypes` to `anypb` at some point in the future?

Comment (alecholmez) [reply pkg/test/resource/v3/resource.go:332]: Sure we can do that

**Comment by delong-coder** [comment]
@alecholmez Is there any changes need to be done here?

**Comment by alecholmez** [comment]
I think this is good to go in, let's make sure the build is all good first, @snowp any objections?

**Comment by alecholmez** [comment]
Oh hmm @delong-coder seems like the build is failing can you take a look?

**Comment by delong-coder** [comment]
> Oh hmm @delong-coder seems like the build is failing can you take a look?

Sure, updated the delta_test.go, please help run the build again, thanks.

**Comment by delong-coder** [comment]
> > Oh hmm @delong-coder seems like the build is failing can you take a look?
> 
> Sure, updated the delta_test.go, please help run the build again, thanks.

@alecholmez

**Comment by alecholmez** [comment]
Cool I think this is good to merge. I won't hold up this logic on a benchmark since that is going to take some time. Thanks @delong-coder

**Review by alecholmez — APPROVED** [review]

**Comment by delong-coder** [comment]
@kyessenov Can I ask what time do we plan to push the new release? Thanks.

**Comment by kyessenov** [comment]
We don't have a time-based policy. If there's demand then it's a matter of compiling the release notes and making sure things are not cut in-the-middle of some work in progress.

**Pull request merged**
